File: code/packages/java/nib-wasm-compiler/src/main/java/com/codingadventures/nibwasmcompiler/NibWasmCompiler.java
package com.codingadventures.nibwasmcompiler;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class NibWasmCompiler {
    public static final String VERSION = "0.1.0";
    private static final int MAX_SOURCE_LENGTH = 1_000_000;
    private static final int MAX_EXPR_NESTING = 256;
    private static final Pattern FUNCTION =
            Pattern.compile("fn\\s+([A-Za-z_][A-Za-z0-9_]*)\\s*\\(([^)]*)\\)\\s*->\\s*u4\\s*\\{\\s*return\\s+([^;]+);\\s*\\}", Pattern.DOTALL);

    private NibWasmCompiler() {}

    public record NibFunction(String name, List<String> params, String expression) {
        public NibFunction {
            params = List.copyOf(params);
            expression = expression.trim();
        }
    }

    public record PackageResult(String source, List<NibFunction> functions, byte[] wasmBytes, Path wasmPath) {
        public PackageResult {
            functions = List.copyOf(functions);
            wasmBytes = wasmBytes.clone();
        }
    }

    public static final class PackageError extends RuntimeException {
        private final String stage;

        public PackageError(String stage, String message) {
            super(message);
            this.stage = stage;
        }

        public String stage() {
            return stage;
        }
    }

    public static PackageResult compileSource(String source) {
        List<NibFunction> functions = parse(source);
        validate(functions);
        return new PackageResult(source, functions, emitModule(functions), null);
    }

    public static PackageResult packSource(String source) {
        return compileSource(source);
    }

    public static PackageResult writeWasmFile(String source, Path path) {
        PackageResult result = compileSource(source);
        try {
            Files.write(path, result.wasmBytes());
        } catch (IOException error) {
            throw new PackageError("write", error.getMessage());
        }
        return new PackageResult(result.source(), result.functions(), result.wasmBytes(), path);
    }

    private static List<NibFunction> parse(String source) {
        if (source.length() > MAX_SOURCE_LENGTH) {
            throw new PackageError("parse", "source exceeds " + MAX_SOURCE_LENGTH + " characters");
        }
        List<NibFunction> functions = new ArrayList<>();
        Matcher matcher = FUNCTION.matcher(source);
        int cursor = 0;
        while (matcher.find()) {
            if (!source.substring(cursor, matcher.start()).trim().isEmpty()) {
                throw new PackageError("parse", "unexpected text before function");
            }
            functions.add(new NibFunction(matcher.group(1), parseParams(matcher.group(2)), matcher.group(3)));
            cursor = matcher.end();
        }
        if (!source.substring(cursor).trim().isEmpty() || functions.isEmpty()) {
            throw new PackageError("parse", "expected one or more Nib functions");
        }
        return functions;
    }

    private static List<String> parseParams(String text) {
        if (text.trim().isEmpty()) {
            return List.of();
        }
        List<String> params = new ArrayList<>();
        for (String piece : text.split(",")) {
            String[] parts = piece.trim().split("\\s*:\\s*");
            if (parts.length != 2 || !"u4".equals(parts[1]) || !parts[0].matches("[A-Za-z_][A-Za-z0-9_]*")) {
                throw new PackageError("parse", "parameters must be `name: u4`");
            }
            params.add(parts[0]);
        }
        return params;
    }

    private static void validate(List<NibFunction> functions) {
        Map<String, NibFunction> byName = new LinkedHashMap<>();
        for (NibFunction function : functions) {
            if (byName.put(function.name(), function) != null) {
                throw new PackageError("validate", "duplicate function `" + function.name() + "`");
            }
        }
        for (NibFunction function : functions) {
            emitExpr(new Section(), function.expression(), byName, paramMap(function), false, 0);
        }
    }

    private static byte[] emitModule(List<NibFunction> functions) {
        Section module = new Section();
        module.write(new byte[] {0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00});

        Section types = new Section();
        types.u32(functions.size());
        for (NibFunction function : functions) {
            types.funcType(function.params().size(), 1);
        }
        module.write(section(1, types.bytes()));

        Section functionSection = new Section();
        functionSection.u32(functions.size());
        for (int index = 0; index < functions.size(); index++) {
            functionSection.u32(index);
        }
        module.write(section(3, functionSection.bytes()));

        Section exports = new Section();
        exports.u32(functions.size());
        for (int index = 0; index < functions.size(); index++) {
            exports.export(functions.get(index).name(), 0x00, index);
        }
        module.write(section(7, exports.bytes()));

        Map<String, NibFunction> byName = new LinkedHashMap<>();
        for (NibFunction function : functions) {
            byName.put(function.name(), function);
        }
        Section code = new Section();
        code.u32(functions.size());
        for (NibFunction function : functions) {
            Section body = new Section();
            body.u32(0);
            emitExpr(body, function.expression(), byName, paramMap(function), true, 0);
            body.write(0x0b);
            byte[] bytes = body.bytes();
            code.u32(bytes.length);
            code.write(bytes);
        }
        module.write(section(10, code.bytes()));
        return module.bytes();
    }

    private static void emitExpr(
            Section out,
            String expression,
            Map<String, NibFunction> functions,
            Map<String, Integer> params,
            boolean emit,
            int depth
    ) {
        if (depth > MAX_EXPR_NESTING) {
            throw new PackageError("validate", "expression nesting exceeds " + MAX_EXPR_NESTING);
        }
        List<String> addParts = splitTopLevel(expression, "+%");
        if (addParts.size() > 1) {
            emitExpr(out, addParts.get(0), functions, params, emit, depth + 1);
            for (int index = 1; index < addParts.size(); index++) {
                emitExpr(out, addParts.get(index), functions, params, emit, depth + 1);
                if (emit) {
                    out.write(0x6a);
                    out.i32(15);
                    out.write(0x71);
                }
            }
            return;
        }
        String trimmed = expression.trim();
        if (trimmed.matches("\\d+")) {
            int value = Integer.parseInt(trimmed);
            if (value < 0 || value > 15) {
                throw new PackageError("validate", "u4 literal out of range: " + value);
            }
            if (emit) {
                out.i32(value);
            }
            return;
        }
        Matcher call = Pattern.compile("([A-Za-z_][A-Za-z0-9_]*)\\s*\\((.*)\\)").matcher(trimmed);
        if (call.matches()) {
            NibFunction target = functions.get(call.group(1));
            if (target == null) {
                throw new PackageError("validate", "unknown function `" + call.group(1) + "`");
            }
            List<String> args = splitArgs(call.group(2));
            if (args.size() != target.params().size()) {
                throw new PackageError("validate", "wrong arity for `" + target.name() + "`");
            }
            for (String arg : args) {
                emitExpr(out, arg, functions, params, emit, depth + 1);
            }
            if (emit) {
                out.write(0x10);
                out.u32(new ArrayList<>(functions.keySet()).indexOf(target.name()));
            }
            return;
        }
        Integer paramIndex = params.get(trimmed);
        if (paramIndex != null) {
            if (emit) {
                out.write(0x20);
                out.u32(paramIndex);
            }
            return;
        }
        throw new PackageError("validate", "unsupported expression `" + expression + "`");
    }

    private static Map<String, Integer> paramMap(NibFunction function) {
        Map<String, Integer> params = new LinkedHashMap<>();
        for (int index = 0; index < function.params().size(); index++) {
            params.put(function.params().get(index), index);
        }
        return params;
    }

    private static List<String> splitArgs(String text) {
        if (text.trim().isEmpty()) {
            return List.of();
        }
        return splitTopLevel(text, ",");
    }

    private static List<String> splitTopLevel(String text, String delimiter) {
        List<String> parts = new ArrayList<>();
        int depth = 0;
        int start = 0;
        for (int index = 0; index < text.length(); index++) {
            char ch = text.charAt(index);
            if (ch == '(') {
                depth++;
            } else if (ch == ')') {
                depth--;
            } else if (depth == 0 && text.startsWith(delimiter, index)) {
                parts.add(text.substring(start, index).trim());
                start = index + delimiter.length();
                index += delimiter.length() - 1;
            }
        }
        parts.add(text.substring(start).trim());
        return parts;
    }

    private static byte[] section(int id, byte[] payload) {
        Section out = new Section();
        out.write(id);
        out.u32(payload.length);
        out.write(payload);
        return out.bytes();
    }

    private static final class Section {
        private final ByteArrayOutputStream out = new ByteArrayOutputStream();

        void write(int value) {
            out.write(value & 0xff);
        }

        void write(byte[] bytes) {
            out.writeBytes(bytes);
        }

        void i32(int value) {
            write(0x41);
            s32(value);
        }

        void u32(int value) {
            write(encodeUnsigned(value));
        }

        void s32(int value) {
            write(encodeSigned(value));
        }

        void funcType(int paramCount, int resultCount) {
            write(0x60);
            u32(paramCount);
            for (int index = 0; index < paramCount; index++) {
                write(0x7f);
            }
            u32(resultCount);
            for (int index = 0; index < resultCount; index++) {
                write(0x7f);
            }
        }

        void export(String name, int kind, int index) {
            name(name);
            write(kind);
            u32(index);
        }

        void name(String value) {
            byte[] bytes = value.getBytes(java.nio.charset.StandardCharsets.UTF_8);
            u32(bytes.length);
            write(bytes);
        }

        byte[] bytes() {
            return out.toByteArray();
        }
    }

    private static byte[] encodeUnsigned(int value) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        int remaining = value;
        do {
            int b = remaining & 0x7f;
            remaining >>>= 7;
            if (remaining != 0) {
                b |= 0x80;
            }
            out.write(b);
        } while (remaining != 0);
        return out.toByteArray();
    }

    private static byte[] encodeSigned(int value) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        int remaining = value;
        boolean more;
        do {
            int b = remaining & 0x7f;
            remaining >>= 7;
            boolean signBit = (b & 0x40) != 0;
            more = !((remaining == 0 && !signBit) || (remaining == -1 && signBit));
            if (more) {
                b |= 0x80;
            }
            out.write(b);
        } while (more);
        return out.toByteArray();
    }
}

File: code/packages/kotlin/nib-wasm-compiler/src/main/kotlin/com/codingadventures/nibwasmcompiler/NibWasmCompiler.kt
package com.codingadventures.nibwasmcompiler

import java.io.ByteArrayOutputStream
import java.nio.file.Files
import java.nio.file.Path

data class NibFunction(val name: String, val params: List<String>, val expression: String)

data class PackageResult(
    val source: String,
    val functions: List<NibFunction>,
    val wasmBytes: ByteArray,
    val wasmPath: Path? = null,
) {
    override fun equals(other: Any?): Boolean =
        other is PackageResult &&
            source == other.source &&
            functions == other.functions &&
            wasmBytes.contentEquals(other.wasmBytes) &&
            wasmPath == other.wasmPath

    override fun hashCode(): Int = 31 * (31 * source.hashCode() + functions.hashCode()) + wasmBytes.contentHashCode()
}

class PackageError(val stage: String, message: String) : RuntimeException(message)

object NibWasmCompiler {
    const val VERSION = "0.1.0"
    private const val MAX_SOURCE_LENGTH = 1_000_000
    private const val MAX_EXPR_NESTING = 256
    private val functionRegex =
        Regex("""fn\s+([A-Za-z_][A-Za-z0-9_]*)\s*\(([^)]*)\)\s*->\s*u4\s*\{\s*return\s+([^;]+);\s*\}""", RegexOption.DOT_MATCHES_ALL)

    fun compileSource(source: String): PackageResult {
        val functions = parse(source)
        validate(functions)
        return PackageResult(source, functions, emitModule(functions))
    }

    fun packSource(source: String): PackageResult = compileSource(source)

    fun writeWasmFile(source: String, path: Path): PackageResult {
        val result = compileSource(source)
        try {
            Files.write(path, result.wasmBytes)
        } catch (error: java.io.IOException) {
            throw PackageError("write", error.message ?: "write failed")
        }
        return result.copy(wasmPath = path)
    }

    private fun parse(source: String): List<NibFunction> {
        if (source.length > MAX_SOURCE_LENGTH) {
            throw PackageError("parse", "source exceeds $MAX_SOURCE_LENGTH characters")
        }
        val functions = mutableListOf<NibFunction>()
        var cursor = 0
        for (match in functionRegex.findAll(source)) {
            if (source.substring(cursor, match.range.first).trim().isNotEmpty()) {
                throw PackageError("parse", "unexpected text before function")
            }
            functions += NibFunction(match.groupValues[1], parseParams(match.groupValues[2]), match.groupValues[3].trim())
            cursor = match.range.last + 1
        }
        if (source.substring(cursor).trim().isNotEmpty() || functions.isEmpty()) {
            throw PackageError("parse", "expected one or more Nib functions")
        }
        return functions
    }

    private fun parseParams(text: String): List<String> {
        if (text.trim().isEmpty()) return emptyList()
        return text.split(",").map { piece ->
            val parts = piece.trim().split(Regex("""\s*:\s*"""))
            if (parts.size != 2 || parts[1] != "u4" || !Regex("""[A-Za-z_][A-Za-z0-9_]*""").matches(parts[0])) {
                throw PackageError("parse", "parameters must be `name: u4`")
            }
            parts[0]
        }
    }

    private fun validate(functions: List<NibFunction>) {
        val byName = functions.associateBy { it.name }
        if (byName.size != functions.size) throw PackageError("validate", "duplicate function")
        for (function in functions) {
            emitExpr(Section(), function.expression, byName, function.params.withIndex().associate { it.value to it.index }, false, 0)
        }
    }

    private fun emitModule(functions: List<NibFunction>): ByteArray {
        val module = Section()
        module.write(byteArrayOf(0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00))

        val types = Section()
        types.u32(functions.size)
        for (function in functions) types.funcType(function.params.size, 1)
        module.write(section(1, types.bytes()))

        val functionSection = Section()
        functionSection.u32(functions.size)
        functions.indices.forEach { functionSection.u32(it) }
        module.write(section(3, functionSection.bytes()))

        val exports = Section()
        exports.u32(functions.size)
        functions.forEachIndexed { index, function -> exports.export(function.name, 0x00, index) }
        module.write(section(7, exports.bytes()))

        val byName = functions.associateBy { it.name }
        val code = Section()
        code.u32(functions.size)
        for (function in functions) {
            val body = Section()
            body.u32(0)
            emitExpr(body, function.expression, byName, function.params.withIndex().associate { it.value to it.index }, true, 0)
            body.write(0x0b)
            val bytes = body.bytes()
            code.u32(bytes.size)
            code.write(bytes)
        }
        module.write(section(10, code.bytes()))
        return module.bytes()
    }

    private fun emitExpr(out: Section, expression: String, functions: Map<String, NibFunction>, params: Map<String, Int>, emit: Boolean, depth: Int) {
        if (depth > MAX_EXPR_NESTING) {
            throw PackageError("validate", "expression nesting exceeds $MAX_EXPR_NESTING")
        }
        val addParts = splitTopLevel(expression, "+%")
        if (addParts.size > 1) {
            emitExpr(out, addParts.first(), functions, params, emit, depth + 1)
            addParts.drop(1).forEach {
                emitExpr(out, it, functions, params, emit, depth + 1)
                if (emit) {
                    out.write(0x6a)
                    out.i32(15)
                    out.write(0x71)
                }
            }
            return
        }
        val trimmed = expression.trim()
        val literal = trimmed.toIntOrNull()
        if (literal != null) {
            if (literal !in 0..15) throw PackageError("validate", "u4 literal out of range: $literal")
            if (emit) out.i32(literal)
            return
        }
        val call = Regex("""([A-Za-z_][A-Za-z0-9_]*)\s*\((.*)\)""").matchEntire(trimmed)
        if (call != null) {
            val target = functions[call.groupValues[1]] ?: throw PackageError("validate", "unknown function `${call.groupValues[1]}`")
            val args = splitArgs(call.groupValues[2])
            if (args.size != target.params.size) throw PackageError("validate", "wrong arity for `${target.name}`")
            args.forEach { emitExpr(out, it, functions, params, emit, depth + 1) }
            if (emit) {
                out.write(0x10)
                out.u32(functions.keys.indexOf(target.name))
            }
            return
        }
        val paramIndex = params[trimmed]
        if (paramIndex != null) {
            if (emit) {
                out.write(0x20)
                out.u32(paramIndex)
            }
            return
        }
        throw PackageError("validate", "unsupported expression `$expression`")
    }

    private fun splitArgs(text: String): List<String> = if (text.trim().isEmpty()) emptyList() else splitTopLevel(text, ",")

    private fun splitTopLevel(text: String, delimiter: String): List<String> {
        val parts = mutableListOf<String>()
        var depth = 0
        var start = 0
        var index = 0
        while (index < text.length) {
            when (text[index]) {
                '(' -> depth++
                ')' -> depth--
            }
            if (depth == 0 && text.startsWith(delimiter, index)) {
                parts += text.substring(start, index).trim()
                start = index + delimiter.length
                index += delimiter.length
                continue
            }
            index++
        }
        parts += text.substring(start).trim()
        return parts
    }

    private fun section(id: Int, payload: ByteArray): ByteArray =
        Section().apply {
            write(id)
            u32(payload.size)
            write(payload)
        }.bytes()
}

private class Section {
    private val out = ByteArrayOutputStream()

    fun write(value: Int) {
        out.write(value and 0xff)
    }

    fun write(bytes: ByteArray) {
        out.write(bytes)
    }

    fun i32(value: Int) {
        write(0x41)
        s32(value)
    }

    fun u32(value: Int) {
        write(encodeUnsigned(value))
    }

    fun s32(value: Int) {
        write(encodeSigned(value))
    }

    fun funcType(paramCount: Int, resultCount: Int) {
        write(0x60)
        u32(paramCount)
        repeat(paramCount) { write(0x7f) }
        u32(resultCount)
        repeat(resultCount) { write(0x7f) }
    }

    fun export(name: String, kind: Int, index: Int) {
        name(name)
        write(kind)
        u32(index)
    }

    private fun name(value: String) {
        val bytes = value.encodeToByteArray()
        u32(bytes.size)
        write(bytes)
    }

    fun bytes(): ByteArray = out.toByteArray()
}

private fun encodeUnsigned(value: Int): ByteArray {
    val out = ByteArrayOutputStream()
    var remaining = value
    do {
        var byte = remaining and 0x7f
        remaining = remaining ushr 7
        if (remaining != 0) byte = byte or 0x80
        out.write(byte)
    } while (remaining != 0)
    return out.toByteArray()
}

private fun encodeSigned(value: Int): ByteArray {
    val out = ByteArrayOutputStream()
    var remaining = value
    var more: Boolean
    do {
        var byte = remaining and 0x7f
        remaining = remaining shr 7
        val signBit = byte and 0x40 != 0
        more = !((remaining == 0 && !signBit) || (remaining == -1 && signBit))
        if (more) byte = byte or 0x80
        out.write(byte)
    } while (more)
    return out.toByteArray()
}
